src/Storages/Statistics/Statistics.cpp

```diff
@@ -15,39 +18,134 @@ namespace ErrorCodes
 {
     extern const int LOGICAL_ERROR;
     extern const int INCORRECT_QUERY;
-    extern const int ILLEGAL_STATISTIC;
 }
 
-void MergeTreeStatisticsFactory::registerCreator(StatisticType stat_type, Creator creator)
+enum StatisticsFileVersion : UInt16
+{
+    V0 = 0,
+};
+
+/// Version / bitmask of statistics / data of statistics /
+
+ColumnStatistics::ColumnStatistics(const ColumnStatisticsDescription & stats_desc_)
+    : stats_desc(stats_desc_), rows(0)
 {
-    if (!creators.emplace(stat_type, std::move(creator)).second)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "MergeTreeStatisticsFactory: the statistic creator type {} is not unique", stat_type);
 }
 
-void MergeTreeStatisticsFactory::registerValidator(StatisticType stat_type, Validator validator)
+void ColumnStatistics::update(const ColumnPtr & column)
 {
-    if (!validators.emplace(stat_type, std::move(validator)).second)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "MergeTreeStatisticsFactory: the statistic validator type {} is not unique", stat_type);
+    rows += column->size();
+    for (const auto & iter : stats)
+    {
+        iter.second->update(column);
+    }
+}
 
+Float64 ColumnStatistics::estimateLess(Float64 val) const
```

Comment:
A more general API which has the advantage of supporting `BETWEEN`s would be

```cpp
Float64 ColumnStatistics::estimateRange(Float64 min, Float64 max) const;
```

`estimateLess` and `estimateMin` will be special cases with `min` or `max` being `inf`.

Reply:
good idea, will do it in the future